components/pd_client/src/util.rs

```diff
@@ -456,42 +507,159 @@ impl PdConnector {
                 }
             }
         }
-        None
+        Err(box_err!(
+            "failed to connect to {:?}",
+            previous.get_members()
+        ))
     }
 
-    pub async fn connect_member(&self, peer: &Member) -> Option<PdClientStub> {
-        for ep in peer.get_client_urls() {
-            if let Ok((client, _)) = self.connect(ep.as_str()).await {
-                info!("connected to PD leader"; "endpoints" => ep);
-                return Some(client);
+    async fn reconnect_pd(
+        &self,
+        members_resp: GetMembersResponse,
+        pre_forwarded_host: String,
+        force: bool,
+        enable_forwarding: bool,
+    ) -> Result<Option<StubTuple>> {
+        let resp = self.load_members(&members_resp).await?;
+        let leader = resp.get_leader();
+        let members = resp.get_members();
+        // Currently we connect to leader directly and there is no member change.
+        // We don't need to connect to PD again.
+        if !force && pre_forwarded_host.is_empty() && resp == members_resp {
+            return Ok(None);
+        }
+        let (res, has_network_error) = self.reconnect_leader(leader).await?;
+        // There are 3 kinds of situations we will return the new client:
+        // 1. the force is true which represents the client is newly created or the original connection has some problem
+        // 2. the previous forwared host is not equal to the one right now which represents the network partition problem to leader may be recovered
+        // 3. the member information of PD has been changed
+        match res {
+            Some((client, forwarded_host)) => {
+                if force || pre_forwarded_host != forwarded_host || resp != members_resp {
+                    return Ok(Some((client, forwarded_host, resp)));
+                } else {
+                    return Ok(None);
+                }
+            }
+            None => {
+                // If the force is false, we could have already forwarded the requests.
+                // We don't need to try forwarding again.
+                if !force {
+                    return Ok(None);
```

Comment:
In the past, error is returned instead.

Reply:
Yes, but for now, I think None is more reasonable?

Reply:
Letting caller know there is a network error seems more reasonable given that `!force` is also expected to reconnecting leader.